### PR TITLE
Add a synonyms feature to support a user defined synonym filter

### PR DESCRIPTION
**Pull request by christianc1**
### Description of the Change

**Work in Progress**

A EP Core Feature to add synonym support to ElasticPress.  Synonyms are stored as post content of a post in a private post type, though the edit UI is decoupled from the standard post edit screen.  Synonym lists follow the Solr spec defined [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html#_solr_synonyms).
 
### Alternate Designs

Needs UX/Visual Design.

### Benefits

Users can define synonyms or alternate spellings for their most popular search terms from the UI.

### Possible Drawbacks

Large synonym filters can affect elasticsearch query performance.

### Verification Process

Unverified.  Work in progress, looking for 👍 's to proceed beyond proof of concept.

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### Applicable Issues

#218 #1791

### CODE REVIEW

File: includes/classes/Feature/Synonyms/Synonyms.php [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+	/**
+	 * Initialize feature setting it's config
+	 *
+	 * @since  3.4
+	 */
+	public function __construct() {
+		$this->slug                     = 'synonyms';
+		$this->title                    = esc_html__( 'Synonyms', 'elasticpress' );
+		$this->requires_install_reindex = false;
```
Comment (felipeelia) [inline includes/classes/Feature/Synonyms/Synonyms.php:45]: As we are changing the mapping, shouldn't this be true? In my tests locally the feature just works after reindexing.

File: includes/classes/Feature/Synonyms/Synonyms.php [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+	 */
+	public function register_post_type() {
+		$args = [
+			'description'        => esc_html__( 'Elasticsearch Synonyms', 'elasticpress' ),
+			'public'             => false,
+			'publicly_queryable' => false,
+			'show_ui'            => false,
+			'show_in_menu'       => false,
+			'query_var'          => true,
+			'rewrite'            => [ 'slug' => 'ep-synonyms' ],
```
Comment (felipeelia) [inline includes/classes/Feature/Synonyms/Synonyms.php:299]: Sorry if I'm missing something obvious but do we need rewrite for this CPT?

File: includes/classes/Feature/Synonyms/Synonyms.php [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+	 * @return void
+	 */
+	public function output_feature_box_long() {
+		?>
+		<p><?php esc_html_e( 'Create a custom synonym filter to allow ElasticPress to match alternative spellings or synonyms of your most popular search terms.', 'elasticpress' ); ?></p>
+		<?php
+	}
+
+	/**
+	 * Adds search synonyms.
```
Comment (felipeelia) [inline includes/classes/Feature/Synonyms/Synonyms.php:112]: Just to keep consistency, this should be "Add search synonyms.", right?

File: includes/classes/Feature/Synonyms/Synonyms.php (hunk 1 above)
Comment (christianc1) [reply includes/classes/Feature/Synonyms/Synonyms.php:45]: I'm not sure, and it's one of the things I think is worthy of a conversation.  Upon activating the feature, I add an annotated synonym file which is only really helpful as a starting point to help users understand the syntax for the file.  (It's the same as the example as the Elasticsearch and Solr docs.

Re-indexing on install would actually put this initial mapping to ES, which I'm not sure is the right behavior.  There is the case where users deactivate then re-activate the feature after adding their own synonyms.  In that case, it makes sense to put that mapping on reactivation.

In an ideal world, I think we put an updated mapping to ES when a user saves the synonym file, or provide the reindex button on the header bar on the synonym page.

File: includes/classes/Feature/Synonyms/Synonyms.php (hunk 3 above)
Comment (christianc1) [reply includes/classes/Feature/Synonyms/Synonyms.php]: updated

File: includes/classes/Feature/Synonyms/Synonyms.php (hunk 2 above)
Comment (christianc1) [reply includes/classes/Feature/Synonyms/Synonyms.php]: removed in 4c73c9c

**Comment by christianc1** [comment]
Update admin UI per @helen's draft.

[](url)

**Comment by christianc1** [comment]
@helen This is ready for any frontend tweaks.  Additionally mentioned elsewhere, one of the editor mode toggle buttons should be removed, whichever you want.  Editor mode preference is saved in the feature settings so it persists (globally).  We could save to user meta but I think that's a bit overkill.

**Pull request merged**